app/src/main/java/com/google/android/sambadocumentsprovider/mount/SectionedAdapter.java

```diff
@@ -0,0 +1,244 @@
+
+  public abstract String getNameForRow(int sectionIndex, int rowIndex);
+  public abstract int getIconForRow(int sectionIndex, int rowIndex);
+
+  @Override
+  public int getCount() {
+    return mFlatRows.size() + (mIsLoadingData ? 1 : 0);
+  }
+
+  @Override
+  public FlatRow getItem(int position) {
+    if (position == mFlatRows.size() && mIsLoadingData) {
+      return null;
+    }
+
+    return mFlatRows.get(position);
+  }
+
+  @Override
+  public long getItemId(int position) {
+    return position;
+  }
+
+  public @Type int getItemViewType(int position) {
+    if (position == mFlatRows.size()) {
+      return LOADING_ROW;
+    }
+
+    return mFlatRows.get(position).isSectionHeader() ? SECTION_HEADER : REGULAR_ROW;
+  }
+
+  @Override
+  public View getView(int position, View reuseView, ViewGroup parent) {
+    int viewType = getItemViewType(position);
+
+    if (reuseView == null) {
+      LayoutInflater inflater = LayoutInflater.from(parent.getContext());
+
+      switch (viewType) {
+        case LOADING_ROW:
+          reuseView = inflater.inflate(R.layout.browsing_progress_item_layout, null);
+          break;
+        case SECTION_HEADER:
+        case REGULAR_ROW:
+          reuseView = inflater.inflate(R.layout.browsing_list_item_layout, null);
+          break;
+        default:
+          // Should never happen.
+          throw new RuntimeException("Unknown view type: " + viewType);
+      }
+
+      reuseView.setTag(new ViewHolder(reuseView));
+    }
+
+    ViewHolder holder = (ViewHolder) reuseView.getTag();
+
+    if (viewType == LOADING_ROW) {
+      return reuseView;
+    }
+
+    holder.bindView(mFlatRows.get(position));
+
+    return reuseView;
+  }
+
+  @Override
+  public boolean isEnabled(int position) {
+    if (position == mFlatRows.size() && mIsLoadingData) {
+      return false;
+    }
+
+    return !mFlatRows.get(position).isSectionHeader();
+  }
+
+  @Override
+  public void notifyDataSetChanged() {
+    flattenData();
+
+    super.notifyDataSetChanged();
+  }
+
+  private void flattenData() {
+    int rowsCount = 0;
+
+    for (int i = 0; i < getSectionCount(); i++) {
+      rowsCount = addFlatRow(getNameForSection(i), getIconForSection(i), true, rowsCount);
+
+      for (int j = 0; j < getSizeForSection(i); j++) {
+        rowsCount = addFlatRow(getNameForRow(i, j), getIconForRow(i, j), false, rowsCount);
+      }
+    }
+  }
+
+  private int addFlatRow(String name, int iconId, boolean isSectionHeader, int position) {
+    FlatRow newRow = FlatRow.create(
+            name,
+            iconId,
+            isSectionHeader,
+            position < mFlatRows.size() ? mFlatRows.get(position) : null);
+
+
+    if (position < mFlatRows.size()) {
+      mFlatRows.set(position, newRow);
+    } else {
+      mFlatRows.add(newRow);
+    }
+
+    return position + 1;
+  }
+
+  protected boolean isLoadingData() {
+    return mIsLoadingData;
+  }
+
+  private static class ViewHolder {
```

Comment:
Actually we can make this adapter more generic by taking view holder for each section & each row.

But we're not for now aiming at anything generic so I'll leave it up to you.